### PR TITLE
Test that `all.css` contains a valid CSS selector

### PR DESCRIPTION
**Pull request by MaxDesiatov**
Verifies that generated `all.css` contains the `:root` string on CI. It doesn't do any actual CSS parsing to verify its full validity, only does a simple `grep` check, but I hope this will be enough to prevent issues like #198 in the future.

### CODE REVIEW

**Comment by mattt** [comment]
Thanks for getting started on this, @MaxDesiatov. I agree with your instinct to start with a shell script for this. Given that the primary offering of `swift-doc` is the CLI, do you think that it'd make sense to invest in a formal Swift test target that exercises this and other behavior of the executable?

**Comment by MaxDesiatov** [comment]
Sure, let's do that. I'll update the PR accordingly 👍

**Comment by MaxDesiatov** [comment]
@mattt ready for review 🙂

**Comment by mattt** [comment]
@MaxDesiatov Nice! Taking a look now.

File: Tests/EndToEndTests/EndToEndTests.swift [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    let process = Process()
+    process.executableURL = productsDirectory.appendingPathComponent("swift-doc")
+    process.arguments = ["generate", "--module-name", "SwiftDoc", "--format", "html", "Sources"]
+
+    let pipe = Pipe()
+    process.standardOutput = pipe
+
+    try process.run()
+    process.waitUntilExit()
```
Comment (mattt) [inline Tests/EndToEndTests/EndToEndTests.swift:19]:
```suggestion
    let process = Process()
    process.executableURL = productsDirectory.appendingPathComponent("swift-doc")
    process.arguments = ["generate", "--module-name", "SwiftDoc", "--format", "html", "Sources"]

    let pipe = Pipe()
    process.standardOutput = pipe

    if #available(OSX 10.13, *) {
      try process.run()
    } else {
      process.launch()
    }
    process.waitUntilExit()
```

Comment (mattt) [reply Tests/EndToEndTests/EndToEndTests.swift:19]: As far as I can tell, the only macOS 10.13 API here is `process.run`, correct? If so, could we tailor the availability check to conditionally use `process.launch` instead where `run` isn't available?

File: Tests/EndToEndTests/EndToEndTests.swift [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      .deletingLastPathComponent()
+      .deletingLastPathComponent()
+      .appendingPathComponent("documentation")
+      .appendingPathComponent("all.css")
+    let cssData = try Data(contentsOf: cssPath)
+    guard let css = String(data: cssData, encoding: .utf8) else {
+      return XCTFail("Failed to decode a UTF-8 string from `cssData` in \(#function)")
+    }
+
+    XCTAssertTrue(css.starts(with: ":root"))
```
Comment (mattt) [inline Tests/EndToEndTests/EndToEndTests.swift:36]:
```suggestion
    XCTAssertTrue(css.contains(":root"))
```

Comment (mattt) [reply Tests/EndToEndTests/EndToEndTests.swift:36]: I could see this assumption breaking unexpectedly, for example if we were to add a header comment or the CSS preprocessor tool decided to reorder or remove the rules. `contains` would be a suboptimal choice if we were worried about performance, but I think this is fine here.

**Review by mattt — CHANGES_REQUESTED** [review]
A few suggested changes for your consideration. Would it be alright if I iterate on this some more this morning, and add some more test cases (e.g. to handle CSS and HTML formats)?

**Comment by MaxDesiatov** [comment]
Sure, please do!

File: Tests/EndToEndTests/EndToEndTests.swift [hunk 3]
```diff
@@ -4,18 +4,18 @@ import XCTest
...
-
     let process = Process()
     process.executableURL = productsDirectory.appendingPathComponent("swift-doc")
     process.arguments = ["generate", "--module-name", "SwiftDoc", "--format", "html", "Sources"]
 
     let pipe = Pipe()
     process.standardOutput = pipe
 
-    try process.run()
+    if #available(OSX 10.13, *) {
```
Comment (mattt) [inline Tests/EndToEndTests/EndToEndTests.swift:13]: Looks like I missed another macOS 10.13 change: https://github.com/SwiftDocOrg/swift-doc/runs/1215042760?check_suite_focus=true#step:5:67

Making that change now.

File: Changelog.md [hunk 4]
```diff
@@ -9,8 +9,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
-- Added an end-to-end test for validity of generated `all.css` file.
-  #199 by @MaxDesiatov.
+- Added end-to-end tests for generate command-line interface.
```
Comment (MaxDesiatov) [inline Changelog.md:12]:
```suggestion
- Added end-to-end tests for `generate` command-line interface.
```

**Review by mattt — APPROVED** [review]

File: Changelog.md (hunk 4 above)
Comment (mattt) [reply Changelog.md]: On second thought, better to remove mention of `generate` entirely, since we're now testing all subcommands.

**Comment by mattt** [comment]
@MaxDesiatov Any last changes before this gets merged?

**Comment by MaxDesiatov** [comment]
LGTM, thank you for finishing this up!

**Comment by MaxDesiatov** [comment]
Can I merge this now? Do you prefer merge commits or linear history by the way?

**Comment by mattt** [comment]
@MaxDesiatov I like to squash and merge PRs. Feel free to do the honors of hitting that green button 😄 (and thanks again for your help with this!)

**Pull request merged**